Various minor crash/etc fixes

insignification opened the pull request:
This includes pull request #21 and #24 and the following fixes (with tests):
- Fixes for jumping out of try/with/except/finally statements (esp. in pypy and now-supported Python 3.8)

It also includes the following non-fix enhancements:
- Cache patched code blocks to avoid unnecessary patching when @with_goto is applied on a nested function.
- Add warnings if something goes wrong for easier debugging. (If you get a warning when applying with_goto, there's a very good chance the resulting function will misbehave or crash)

All locally tested on Python 2.6/2.7/3.4/3.8 & pypy2.7&3.6

insignification committed:
Properly fix `Jump out of too many nested blocks` error

(overrides previous fix, as merging them would be too complex and
unneeded)

Commit:
Rename new SyntaxError to be more descriptive in case ever encountered.

I was not able to get it to happen for pretty huge functions (much much
larger than the ones in the test), though, so there's a good chance it's
unreachable.

Commit:
Refactor previous changes to avoid duplication & fix safety issues

(Candidate for moving to 'fix' branch?)

Commit:
Added python 3.8 support

Added 4 tests, 3 currently disabled (pre-existing issues, especially on
pypy).

Commit:
Fix test_jump_out_of_try_block_and_survive.

Add new test_jump_out_of_try_block_and_live

Commit:
Avoid patching code again and again in nested functions

Uses a weak dictionary just in case code objects can be gc'd (can they?
maybe on some implementations?)

Commit:
Fix previous change for python2.6 (doesn't support Code weakrefs)

Commit:
Fix with blocks (add tests)

add warnings for easier debugging (for peace of heart - they weren't
currently triggered by any tests)

Commit:
Support jumping out of except & finally blocks!

Commit:
Add combined try/catch/finally test and remove unneccessary op

(later fix could go into fix branch as well)

Commit:
Fixed two dumb bugs that cancelled each other out in most cases

(No testcase in this branch, as it'd be tricky to think of one. master
branch has testcases becaues it does dis.findlabels)

Commit:
Fix jumping out of with blocks in py26 (and added test)

Plus added extra test from master

snoack commented:
Thanks, very much appreciated! But reviewing this huge PR is difficult. Let's land #21 first separately, and if you don't mind can you move the changes that are unrelated of supporting Python 3.8 to separate pull requests?

Also can you update `tox.ini`, `.travis.yml` and `README.md` in order to cover Python 3.8?

Commit:
Style & test updates (other updates to commit in following commits)

Commit:
Refactor the ops sizing/writing to avoid code duplication

(This was originally in the second PR, but since you want to take this
one first, I think it's a good commit to add here, since it's relevant
to this change and would make diffing against future PRs easier)

Commit:
Updated style & added py38 to tox/travis/readme

Commit:
fix line endings snafu (you'll squash anyway, so I'm keeping history)

Commit:
Remove the commented out tests (They'll come back in the next PR).

(The new test helps test functionality that needs special attention at
least on py38)

insignification committed:
Merge branch 'master' into py38

Commit:
Update merge

insignification commented:
I've split this PR into py38 and non-py38.
Splitting the PR further would be too much work, though, and it contains mostly related fixes (+ one unrelated minor change that should hopefully be easy to review together with the rest)

Commit:
Added while / while true tests & fix code to make them pass

insignification committed:
Merge branch 'master' into fix2

snoack commented:
I didn't do a full review yet (as this patch is difficult to review until rebased/merged against `master`) but I already have a couple remarks:

1. I'm not sure if I like the caching. It seems like a premature optimization, and I cannot think of a common scenario where the same code is passed through the `goto` decorator more than once.
2. Specifically detecting `__pypy__` seems wrong. I rather stick to feature detection (e.g. compile some code, and inspect the resulting bytecode, like I did in order to detect Python 3.6+ wordcode, see the `_Bytecode` class and how its properties are initialized).

insignification committed:
Merge branch 'master' into fix2

Commit:
Style fixes

Commit:
Added tests that show remaining issues, first part of fixing them.

Commit:
Backport change from 'iter' needed for the newly added tests

(Change makes it so POP_BLOCK is no longer looked at (as it's a poor
indicator in some cases), instead - we look at the target of each block.
Also includes a simplification removing block_exits)

insignification commented:
I've updated the changes per flake8 & your notes in the other PR.

To respond to your questions above:

1) The common scenario that this handles is as follows:
```
def outer_func():
        @with_goto
        def inner_func():
            <do something func with gotos>
        return inner_func
```
Since the decorator runs every time outer_func is called, we would needlessly patch the function's bytecode every time (E.g. imagine outer_func is called in a heavy loop).
This results not only in needless speed loss but also in needless creation of duplicate code objects.
This is why caching is required and is not a premature optimization.
As for  a memory leak - that's what using a weak-keyed dictionary should resolve. (Supported in py2.7+). Plus, code objects usually have infinite scope, so it's not a real concern even in py2.6 (compared with the alternative).

2) Hmm, that could work - I'll give it a try.

Additional notes:

3) I added two "noqa"s - one for the __pypy__ import (will probably be removed per (2)) and one for a test checking the "except:" case (I moved most tests to check the "except Exception:" case, but keeping one good test for the "except:" case is important too.

4) Due to changing the "except" to "except Exception" in the testcases, I ended up finding a new issue. To resolve it, I made some additional code changes, including "backporting" a relatively major change from the next PR - code no longer relies on POP_BLOCK but instead looks at the jump argument supplied in the SETUP_* opcodes, which seems more reliable in several cases.
(To be clear, the "backported" changes are relevant to *this* PR - the fixing of various crashes/etc in existing supported python-goto functionality)

5) Let me know if you want me to make the following improvement to the code:
   Use namedtuple instead of the brittle _,_,x=y and y[0] approaches. (Supported in py26+ so should be good)

EDIT: I'll also do a refactor of _find_labels_and_gotos to make it use a class for the block_stack management, as it's currently quite brittle (especially due to the lack of python3.x's nonlocal)
EDIT EDIT: Done.

Commit:
Refactor _find_labels_and_gotos to use a class instead of nested funcs

Commit:
Avoid __pypy__ import and instead always call END_FINALLY when possible

(Not possible in 3.8 and as comment says - it's not clear how pypy and
py38 will reconcile their differences there)

Commit:
Go a step further and always call END_FINALLY - make it possible in 3.8

While this isn't strictly required, making it happen requires fully
handling the has_setup_except variation (replacing SETUP_FINALLY with
SETUP_EXCEPT where appropriate) so it's a good thing to have in the repo
for the future

Commit:
flakate

insignification commented:
About (2), while it's not possible to (normally?) detect this at runtime (since the generated opcodes are the same - their semantic is different), I realize I could largely remove the feature check altogether and always write the END_FINALLY when exiting finally/with blocks (previously - was only written for pypy).

In fact, one could well say that not calling END_FINALLY in such cases is a bug, just one that non-pypy pythons don't care about (of course, one could say the same about adding goto to python in general, but let's not go there :P)

Let me know if you'd rather have the __pypy__ feature check over writing 2 extra opcodes (BEGIN_FINALLY/equivalent & END_FINALLY) in all relevant cases. The extra 2 opcodes shouldn't be a problem now that we can write how much we want.

Also, I went ahead a bit further and made some changes to fully handle the _BYTECODE.has_begin_finally flag by translating SETUP_FINALLY to SETUP_EXECPT once detected as well. This allowed me to remove the pypy feature check entirely (as opposed to under the !has_begin_finally feature check).
It's not strictly required, but fits better with the "you should call END_FINALLY after finally/with blocks" narrative and will be helpful if we ever want to add other changes that rely on SETUP_EXCEPT.

This PR is now ready from my perspective. Let me know if you want any changes.

Commit:
Update comment

Commit:
Minor code change
